cfme/middleware/server.py

```diff
@@ -46,12 +62,44 @@ def __init__(self, name, provider=None, **kwargs):
 
     @classmethod
     def servers(cls, provider):
-        sel.force_navigate('middleware_servers')
-        deployments = []
+        servers = []
+        if provider:
+            # if provider instance is provided try to navigate provider's servers page
+            # if no servers are registered in provider, returns empty list
+            if provider.load_all_provider_servers():
+                return servers
+        else:
+            # if provider instance is not provided then navigates  to all servers page
+            sel.force_navigate('middleware_servers')
         for row in list_tbl.rows():
-            deployments.append(MiddlewareServer(
-                provider=provider, name=row.server_name.text, product=row.product.text))
-        return deployments
+            servers.append(MiddlewareServer(
+                name=row.server_name.text, id=row.feed.text,
+                product=row.product.text, provider=row.provider.text))
+        return servers
+
+    @classmethod
+    def headers(cls):
+        headers = [sel.text(hdr).encode("utf-8") for hdr in list_tbl.headers if hdr.text]
```

Comment:
sel.force_navigate('middleware_servers') requires list_tbl._update_cache() otherwise list_tbl.headers fails.
"list_tbl._update_cache()" is recognized as a problematic code for bot.